dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +247,365 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
                     }
-                }
-            ) };
 
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::ComActivator))
-            {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
+                    auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                  
+                    LOG_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get()));
+
+                    LOG_IF_FAILED(notificationsLongRunningPlatform->UnregisterFullTrustApplication(m_processName.get()));
+                }) };
+
+                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                wil::unique_cotaskmem_string unpackagedAppUserModelId;
+                // Apps treated as unpackaged need to call RegisterFullTrustApplication and register with the LRP
+                THROW_IF_FAILED(notificationPlatform->RegisterFullTrustApplication(m_processName.get(), GUID_NULL, &unpackagedAppUserModelId));
 
+                // Register a sink for the application to receive foreground raw notifications
+                RegisterForegroundSinkHelper();
                 {
-                    auto lock{ s_activatorInfoLock.lock_shared() };
-                    THROW_HR_IF_MSG(E_INVALIDARG, s_comActivatorRegistration, "ComActivator already registered.");
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = true;
                 }
 
-                GetWaitHandleForArgs().create();
-                GetComServerClsid() = taskClsid;
-                
-                THROW_IF_FAILED(::CoRegisterClassObject(
-                    taskClsid,
-                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
-                    CLSCTX_LOCAL_SERVER,
-                    REGCLS_MULTIPLEUSE,
-                    &s_comActivatorRegistration));
-            }
+                // m_registeredClsid is set to GUID_NULL for unpackaged applications
+                THROW_IF_FAILED(notificationPlatform->RegisterLongRunningActivatorWithClsid(m_processName.get(), m_registeredClsid));
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonBackgroundRegistration = true;
+                }
 
-            if (builder)
-            {
-                registeredTaskFromBuilder = builder.Register();
-            }
+                // Register a COM object for the PushNotificationLongRunningProcess to CoCreate in background activation scenarios
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
```

Comment:
lock?

Reply:
There is a lock right before CoRegisterClassObject